src/vote_graph.rs

```diff
@@ -0,0 +1,515 @@
+	///
+	/// The GHOST (hash, number) returned will be the block with highest number for which the
+	/// cumulative votes of descendents and itself causes the closure to evaluate to true.
+	///
+	/// This assumes that the evaluation closure is one which returns true for at most a single
+	/// descendent of a block, in that only one fork of a block can be "heavy"
+	/// enough to trigger the threshold.
+	pub fn find_ghost<'a, F>(&'a self, current_best: Option<(H, usize)>, condition: F) -> Option<(H, usize)>
+		where F: Fn(&V) -> bool
+	{
+		let entries = &self.entries;
+		let get_node = |hash: &_| -> &'a _ {
+			entries.get(hash)
+				.expect("node either base or referenced by other in graph; qed")
+		};
+
+		let mut node_key = current_best
+			.and_then(|(hash, number)| match self.find_containing_nodes(hash.clone(), number) {
+				None => Some(hash),
+				Some(ref x) if !x.is_empty() => Some(x[0].clone()),
+				Some(_) => None,
+			})
+			.and_then(|hash| get_node(&hash).ancestor_node())
+			.unwrap_or_else(|| self.base.clone());
+
+		let mut active_node = get_node(&node_key);
+
+		if !condition(&active_node.cumulative_vote) { return None }
+
+		// breadth-first search starting from this node.
+		loop {
+			let next_descendent = active_node.descendents
+				.iter()
+				.map(|d| (d.clone(), get_node(d)))
+				.filter(|&(_, ref node)| condition(&node.cumulative_vote))
+				.next();
+
+			match next_descendent {
+				Some((key, node)) => {
+					node_key = key;
+					active_node = node;
+				}
+				None => break,
+			}
+		}
+
+		// active_node and node_key now correspond to the head node which has enough cumulative votes.
+		// we have a frontier of vote-nodes which individually don't have enough votes
+		// to pass the threshold but some subset of them join either at `active_node` or at some
+		// descendent block of it, giving that block sufficient votes.
+		Some(self.ghost_find_merge_point(node_key, active_node, condition))
+	}
+
+	// given a key, node pair (which must correspond), assuming this node fulfills the condition,
+	// this function will find the highest point at which its descendents merge, which may be the
+	// node itself.
+	fn ghost_find_merge_point<'a, F>(&'a self, mut node_key: H, mut active_node: &'a Entry<H, V>, condition: F) -> (H, usize)
+		where F: Fn(&V) -> bool
+	{
+		let mut descendent_nodes: Vec<_> = active_node.descendents.iter()
+			.map(|h| self.entries.get(h).expect("descendents always present in node storage; qed"))
+			.collect();
+
+		let base_number = active_node.number;
+		let (mut best_hash, mut best_number) = (node_key, active_node.number);
+		let mut descendent_blocks = Vec::with_capacity(descendent_nodes.len());
+
+		// TODO: for long ranges of blocks this could get inefficient
+		for offset in 1usize.. {
+			let mut new_best = None;
+			for d_node in descendent_nodes.iter() {
+				if let Some(d_block) = d_node.ancestor_block(base_number + offset) {
```

Comment:
We return `None` if the block is past the parent vote-node of `d_node`. It's ok to ignore it since the parent vote-node of `d_node` should eventually be reached (if needed) as the offset grows.

Reply:
the parent vote-node of `d_node` will always be the GHOST vote-node. `ancestor_block` will return `None` when we have reached the block corresponding to `d_node`, at which point we will want to prune anyway.